### PR TITLE
Add image related LCM types

### PR DESCRIPTION
**Pull request by kunimatsu-tri**
These lcmtypes are meant to replace [bot_core_images_t.lcm](https://github.com/openhumanoids/bot_core_lcmtypes/blob/master/lcmtypes/bot_core_images_t.lcm) and [bot_core_image_t.lcm](https://github.com/openhumanoids/bot_core_lcmtypes/blob/master/lcmtypes/bot_core_image_t.lcm).

Let me summarize the major differences I made:

**images_t** (now **image_array_t** to clearly differentiate the name from image_t)
1. Dropped `image_types` since `DISPARITY`, `MASK` and `DEPTH` related consts moved to image_t now.  Also, `LEFT` and `RIGHT` are hardware specific information and those can be specified in `header.frame_name`(see below)  inside `image_t` now.

**image_t**

1.  Added `header` so that you can specify at what frame this image is obtained. It's possible for different images to have the same `frame_name` in the `header`, e.g. an RGB image and a depth image registered onto the RGB image have the same frame name.

2. Dropped `image_metadata_t` simply because I couldn't find any particular use cases.

3. Split `pixelformat` into `pixel_format`, `channel_type`, `compression_method` and `bigendian` to reduce redundancy and clean things up.

4. Dropped some consts in `pixelformat`, e.g. UYVY, YUYV, etc. simply because I don't see any use cases yet. I'm OK to add new const with as-needed basis.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/lcmtypes/13)

### CODE REVIEW

**Comment by kunimatsu-tri** [comment]
+@patmarion +@ericcousineau-tri feature review please.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-Kl-NZlECgT-64wdy9w6:bafatqf)*

**Comment by EricCousineau-TRI** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[lcmtypes/image_array_t.lcm, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-VuLhxWiYCpXFEJ1t:-Kl-VuLibBEGt8uIhFGl:b-hzkixg) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_array_t.lcm#L6)):*
> ```
> struct image_array_t {
>   // The timestamp and the frame name.
>   // The timestamp holds when this data is packed. It's convinient to store
> ```

nit spelling, `s/convinient/convenient/`.

---

*[lcmtypes/image_array_t.lcm, line 7 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-W0bDbWWlO0K3xRoh:-Kl-W0bDbWWlO0K3xRoi:b9kqrz) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_array_t.lcm#L7)):*
> ```
>   // The timestamp and the frame name.
>   // The timestamp holds when this data is packed. It's convinient to store
>   // the latest timestamp among the timestamps in `images` since it is possible
> ```

nit Reword, change `images` to `image_t` (or include the type as well), and change `among` to `alongside` to make it a little clearer. (`among` to me indicates mixing of order).

---

*[lcmtypes/image_t.lcm, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-V8IoFIJ3xqIJkvon:-Kl-V8IpVsz50o16ZjRP:baxqahq) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_t.lcm#L35)):*
> ```
>   int8_t compression_method;
> 
>   // enum for pixel_format. Add other formats as needed.
> ```

I think the `Add other formats as needed` is redundant. It may be more helpful to indicate these are only uniformly-sized channels, whose sizes are multiples of bytes, and does not involve any bit-packing like 5-6-5 (16-bit pixels) or what not.

---

*[lcmtypes/image_t.lcm, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-VOlAhAr5NkNLLsHt:-Kl-VOlB29-hHsxPWMCF:bov8j0i) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_t.lcm#L51)):*
> ```
>   const int8_t PIXEL_FORMAT_INVALID    = -1;
> 
>   // enum for channel_type. Add other types as needed.
> ```

Same as above for `Add other types as needed`. I would remove this, and refer to the above comment in code (something like "please see the comments for `pixel_format`")

---

*[lcmtypes/image_t.lcm, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-VY9tdBbxxJI6rz2p:-Kl-VY9tdBbxxJI6rz2q:b2nreh6) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_t.lcm#L62)):*
> ```
>   const int8_t CHANNEL_TYPE_INVALID = -1;
> 
>   // enum for compression_method. Add other methods as needed.
> ```

Same as above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13)*

**Comment by patmarion** [comment]
There's no question that this is an improvement, you have fixed the flaws in the previous message designs.  We discussed these flaws at the beginning of the DRC, but ultimately decided to continue with image_t for compatibility with existing tools.

**Comment by kunimatsu-tri** [comment]
Thanks @patmarion!

Do you remember what was the main blocker at that time?

We may be able to find a solution for that this time since we don't have hard deadline like DRC.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-Kl-aDGtZoeg2sd4OOKI:b-1alsfo)*

**Comment by patmarion** [comment]
I think the choice was just to select a compatible message type rather than make a new one.  As drake users you guys are not as integrated with MIT projects, but for the DRC we wanted to leverage existing code developed by research groups that used the libbot types.

**Comment by kunimatsu-tri** [comment]
If "to select a compatible message type" would be always correct answer, then it would mean there is no way to change the interface forever, but it's not true. (I'm sure you're not saying it's true, but just that was the choice at that moment.)

I can keep backward compatibility with libbot types if you really need while I introduce these new types.

One way to do that is to write an "adapter" script that subscribes a lcm message, converts to another lcm message, and publishes it.

Or, if it would be just a matter of time for you to change director's code to be compatible to these new types, I'm OK to
 work on that too.

I'm paying my attention to be compatible with director very much since director is awesome and very important tool to Drake. I don't know much about libbot tools yet. So, I want to learn from you at that point.

To summarize, my points are:
1. graduation from bot_core_lcmtypes (at least in Drake) because of the flaws and,
2. backward compatibility with director, spartan and anything that MIT folks care.

I'm open to any ideas to make these things happen.

What would you think?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-Kl-yChUfN-YVQ6CUrcz:b-8r9byf)*

**Comment by kunimatsu-tri** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[lcmtypes/image_array_t.lcm, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-VuLhxWiYCpXFEJ1t:-Kl0IyCLC1eAnD6j4Jct:b4lrs88) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_array_t.lcm#L6)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

nit spelling, `s/convinient/convenient/`.
</blockquote></details>

Oops, thanks! Done.

---

*[lcmtypes/image_array_t.lcm, line 7 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-W0bDbWWlO0K3xRoh:-Kl0NETUWrNN0yKfxutA:b-dlkrej) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_array_t.lcm#L7)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

nit Reword, change `images` to `image_t` (or include the type as well), and change `among` to `alongside` to make it a little clearer. (`among` to me indicates mixing of order).
</blockquote></details>

Let's talk in f2f what I meant here.

---

*[lcmtypes/image_t.lcm, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-V8IoFIJ3xqIJkvon:-Kl-oRO5jvYZaK_tHFCv:b-8sl3jf) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_t.lcm#L35)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

I think the `Add other formats as needed` is redundant. It may be more helpful to indicate these are only uniformly-sized channels, whose sizes are multiples of bytes, and does not involve any bit-packing like 5-6-5 (16-bit pixels) or what not.
</blockquote></details>

I thought about this more. 

I found we don't necessarily mention "only uniformly-sized channels"  here because if we could regard YUV422's all the elements as a unified single channel, there is no problem. 

Namely, you can do something like this:
```
pixel_format = PIXEL_FORMAT_YUV422;
channel_type = CHANNEL_TYPE_UINT16;  // 8 bits for Y, 4 bits for both U and V.
```

Done for deleting "Add other ... ".

---

*[lcmtypes/image_t.lcm, line 51 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-VOlAhAr5NkNLLsHt:-Kl0MIzU42rWjwgLslF7:brwixrj) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_t.lcm#L51)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Same as above for `Add other types as needed`. I would remove this, and refer to the above comment in code (something like "please see the comments for `pixel_format`")
</blockquote></details>

Done (deleted).

---

*[lcmtypes/image_t.lcm, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-VY9tdBbxxJI6rz2p:-Kl0MKWEB4vXdCRHHsAT:brwixrj) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_t.lcm#L62)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

Done (deleted).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13)*

**Comment by patmarion** [comment]
Adding support for both message types to director should be straightforward.  The only file you need to update is ddBotImageQueue.cpp in director.  However, director calls into a DRC-era library (multisense utils) to convert the rgbd images (depth or dispartity) to a pointcloud, and passes the data into that library using images_t as a container rather than passing image buffer data, so that could would need to be replaced.  This code is in need of a refactor, anyway.

https://github.com/RobotLocomotion/director/blob/master/src/app/ddBotImageQueue.cpp#L712

**Comment by kunimatsu-tri** [comment]
Thanks for the pointer.  Where is the code for "multisense_utils" located? I want to take a look.

Probably, instead of  touching multisense_utils part, writing a converter between bot_core/image_t and robotlocomotion/image_t is a good way to go as the first step.

BTW, it seems we both agreed not to break backward compatibility to bot_core/image_t in director while I'll introduce robotlocomotion/image_t, can you LGTM to this PR?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-KlP673soNUo3lJN1kob:b9cu1xx)*

**Comment by kunimatsu-tri** [comment]
*[lcmtypes/image_array_t.lcm, line 7 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-W0bDbWWlO0K3xRoh:-KlPBFoq6Dwz9Q3yZdtA:bqmo6w5) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_array_t.lcm#L7)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Let's talk in f2f what I meant here.
</blockquote></details>

Solved with f2f. 

Since `images` in the comment is quoted with backticks, the meaning is clear and we don't need to change anything here.

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13)*

**Comment by patmarion** [comment]
LGTM, but, given my reduced involvement as an LCM user going forward, I think you should seek a different reviewers for new LCM types.

**Comment by EricCousineau-TRI** [comment]
Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[lcmtypes/image_t.lcm, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-Kl-V8IoFIJ3xqIJkvon:-KlPP6DkCfvMolTHg_3S:b-xu7f53) ([raw file](https://github.com/robotlocomotion/lcmtypes/blob/67c878e936e1b0fa0cca05c734dbe94e143d25b3/lcmtypes/image_t.lcm#L35)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I thought about this more. 

I found we don't necessarily mention "only uniformly-sized channels"  here because if we could regard YUV422's all the elements as a unified single channel, there is no problem. 

Namely, you can do something like this:
```
pixel_format = PIXEL_FORMAT_YUV422;
channel_type = CHANNEL_TYPE_UINT16;  // 8 bits for Y, 4 bits for both U and V.
```

Done for deleting "Add other ... ".

</blockquote></details>

OK. Sounds good!

In a f2f, we discussed that even for non-byte-width channels (e.g., `RGB565`), the pixels would effectively become one channel (e.g., `uint16`), to simplify implementation details. Anything that would need to extract the actual color channels on non-byte boundaries would be done external to this implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13)*

**Comment by EricCousineau-TRI** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-KlPPuEyrRIF8t-08IWn:bnfp4nl)*

**Comment by kunimatsu-tri** [comment]
+@manuelli feature review please.
+@sammy-tri platform review please.

cc @RussTedrake

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-KlP_D_qr7Q2_zK2gzne:b-fmgkg7)*

**Comment by sammy-tri** [comment]
:lgtm:

---

Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-KlPeWM8Spedzbo5_maC:bnfp4nl)*

**Comment by manuelli** [comment]
LGTM

**Comment by kunimatsu-tri** [comment]
-@manuelli since just the way of LGTM differs from reviewable's and that's preventing from merging this PR.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/lcmtypes/13#-:-KlQPBnS9dbr23OKe1ZM:b-se8ifu)*
